### PR TITLE
Simplify translation workflow

### PR DESCRIPTION
**Pull request by FeralChild64**
# Description

- removed old *.txt translation files, translators are now requested to edit *.lng files directly
- adapted all the existing scripts
- fixed update sources script which didn't list obsolete strings in the output
- added a README file with some basic information for the translators

## Related issues

Fixes https://github.com/dosbox-staging/dosbox-staging/issues/2429

# Manual testing

I have checked that normalization scripts still seem to work. I have performed all the steps requested by update sources script and checked that it detects obsolete strings.

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [x] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by Kappa971** [comment]
You should also add "normalize-translation.bat" for Windows and restore the tools folder. If I understand correctly, you are removing the utf-8 txt files, leaving only the lng files (which are the same thing).

**Comment by FeralChild64** [comment]
@Kappa971 Yes, I'm removing the `*.txt` files, DOSBox Staging can now handle even non-normalized translation files, so they are not needed at all. Yes, the `*.bat` file will be back (I see you have adapted it independently, I'll look at it later), I just need to test it.

And I need to find out why _Code analysis / Script linters (push)_ check fails...

File: contrib/resources/translations/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+- [article on Wikipedia](https://en.wikipedia.org/wiki/Printf)
+- [article on cplusplus.com](https://cplusplus.com/reference/cstdio/printf)
+- [article on cppreference.com](https://en.cppreference.com/w/cpp/io/c/fprintf)
+
+### Configuration option descriptions
+
+For example `CONFIG_FULLSCREEN`.
+
+Do not exceed 79 character in a line, or commands like `config -h fullscreen`
+won't be able display the help properly. The limit might be increased to 80
```
Comment (rderooy) [inline contrib/resources/translations/README.md:47]: won't be able to display the help properly, and might wrap the text or display blank lines.

File: contrib/resources/translations/README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+characters in the future, but the implementation is not there yet.
+
+### Command or program help
+
+For example `PROGRAM_MOUNT_HELP_LONG` or `SHELL_CMD_CLS_HELP_LONG`. Sometimes
+the first line is extracted to another string, like `SHELL_CMD_CLS_HELP`.
```
Comment (rderooy) [inline contrib/resources/translations/README.md:53]: the first line of the help message is extracted to a separate string

File: contrib/resources/translations/README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+Do not exceed 80 character in a line, or commands like `cls /?` won't be able to
```
Comment (rderooy) [inline contrib/resources/translations/README.md:55]: Did we not cover this already?

File: contrib/resources/translations/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+display the help properly. Also test that command `help /all` properly extracts
+and displays the shortened command descriptions.
```
Comment (rderooy) [inline contrib/resources/translations/README.md:57]: Also test that the`help /all` command properly extracts and displays the shortened command descriptions.

File: contrib/resources/translations/README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+### Startup screen strings
+
+For example `SHELL_STARTUP_BEGIN`, `SHELL_STARTUP_CGA`.
+
+If you touch them, test the startup screen with `startup_verbosity = high` for all
+the relevant `machine = (...)` settings. It is really easy to make a mistake here.
```
Comment (rderooy) [inline contrib/resources/translations/README.md:64]: If you modify these sections, test the startup screen with `startup_verbosity = high` for all the relevant `machine = (...)` settings. It is really easy to make a mistake here.

File: contrib/resources/translations/README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+# Scripts
+
+## `normalize.sh` (Linux, macOS) or `normalize.bat` (Windows)
+
+UTF-8 can store the same exact text in several different ways. There are some standard
+ones, however, called _normalized forms_. These scripts convert the translation files
+to the NFC-normalized form (recommended by DOSBox Staging, saved by most UTF-8 capable
```
Comment (rderooy) [inline contrib/resources/translations/README.md:72]: and used by default by most UTF-8...

**Review by rderooy — APPROVED** [review]
A few documentation updates

**Comment by Kappa971** [comment]
I tried my `normalize.bat` and it works. I removed `dos2unix` because it is no longer required (it seems that the -o parameter of uconv doesn't change the line break in CRLF). Please update `normalize.bat` and `tools/readme.txt` with what I indicated above (if you want you can test it first).

File: contrib/resources/translations/README.md (hunk 3 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md:55]: No, the limit for config options is still 79 characters.

File: contrib/resources/translations/README.md (hunk 6 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+## Preparing file with English strings
+
+First, you need up to date the file with English strings. The one present in this directory is most likely outdated; to produce the current one do the following:
+
+1. In the configuration file, go to section `[dosbox]` and set`language = en`
+2. Run the most current version of DOSBox Staging
+3. Inside DOSBox execute command `config -wl en.txt` (you can supply host OS path instead of bare `en.txt`)
+
+This will produce file with current strings to translate, use it as a reference.
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
+
+### General translation workflow
+
+## Preparing file with English strings
+
+First, you need up to date the file with English strings. The one present in this directory is most likely outdated; to produce the current one do the following:
+
+1. In the configuration file, go to section `[dosbox]` and set`language = en`
+2. Run the most current version of DOSBox Staging
+3. Inside DOSBox execute command `config -wl en.txt` (you can supply host OS path instead of bare `en.txt`)
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+### Format strings
+
+Many strings contain strange sequences starting from percent sign, like `%s` or `%02d`.
+These are so-called _printf format strings_, used by C/C++ programming languages;
+they are replaced with another string or number, using a given format.
+
+You might need to understand the notation to properly format the output (like tables,
+when the translated headers have different length than in English). Search the web
+to learn more about it, here are few links:
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: Ok, changed.

File: contrib/resources/translations/README.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+- [article on Wikipedia](https://en.wikipedia.org/wiki/Printf)
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
+
+### General translation workflow
+
+## Preparing file with English strings
+
+First, you need up to date the file with English strings. The one present in this directory is most likely outdated; to produce the current one do the following:
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: I actually mean "you need an up to date file with English strings". "produce" to "generate" - changed.

File: contrib/resources/translations/README.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+1. In the configuration file, go to section `[dosbox]` and set`language = en`
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+- In the configuration file, go to section `[dosbox]` and set`language = en`.
+- Run the most current version of DOSBox Staging.
+- Inside DOSBox execute command `config -wl en.txt` (you can supply host OS path
+  instead of bare `en.txt`).
+
+This will produce file with current strings to translate, use it as a reference.
+
+## Updating the translation
+
+To update translation just edit the relevant `*.lng` file with text editor,
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md (hunk 2 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+Depending on your host OS, use the correct script to normalize the output. Although
+DOSBox Staging should be able to handle non-normalized translation files, and
+normally text editors save normalized files already, it is still highly recommended
+to use the script to make sure no problem occurs.
+
+## String-specific tips
+
+### Format strings
+
+Many strings contain strange sequences starting from percent sign, like `%s` or `%02d`.
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: I think this would be the best:

_Many strings contain strangely looking sequences starting from the percentage sign,
like %s or %02d_

The prinf format string special sequences always start with percentage sign.

File: contrib/resources/translations/README.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+These are so-called _printf format strings_, used by C/C++ programming languages;
+they are replaced with another string or number, using a given format.
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: I think _specified format_ is even better.

File: contrib/resources/translations/README.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+- [article on Wikipedia](https://en.wikipedia.org/wiki/Printf)
+- [article on cplusplus.com](https://cplusplus.com/reference/cstdio/printf)
+- [article on cppreference.com](https://en.cppreference.com/w/cpp/io/c/fprintf)
+
+### Configuration option descriptions
+
+For example `CONFIG_FULLSCREEN`.
+
+Do not exceed 79 character in a line, or commands like `config -h fullscreen`
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md (hunk 4 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md (hunk 1 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md (hunk 5 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: I think _If you modify these strings_ is better.

**Comment by Kappa971** [comment]
> In the configuration file, go to the `[dosbox]` section and set`language = en`.
Run the most current version of DOSBox Staging.
Inside DOSBox execute the command `config -wl en.txt` (you can supply a host OS
  path, instead of simply `en.txt`).

Or
In the DOSBox Staging folder run `dosbox -lang en -c "config -wl en.lng" -c "exit"`

I discovered that, unfortunately, DOSBox generates an `en.lng` file (I think you can use _lng_ extension instead of _txt.)_ with CRLF line break, so if someone creates a new translation from that file, they will have a file with CRLF line break.
It appears that `uconv,` with the **-o** parameter, leaves the line break unchanged so if the file previously had a CRLF line break, it will remain CRLF. We need to restore dos2unix 😒 Sorry.

**normalize.bat**
```bat
@ECHO OFF

SET translation_dir=%~dp0
PUSHD "%translation_dir%"

IF NOT EXIST tools\uconv.exe GOTO readme
IF NOT EXIST tools\dos2unix.exe GOTO readme

ECHO In directory %translation_dir%:

FOR %%f IN (*.lng) DO (
	CALL :nconv_t "%%f"
)

PAUSE
EXIT /B %ERRORLEVEL%

REM Normalization - for UTF-8 format locale

:nconv_t
ECHO Normalizing file %~1
tools\uconv -f UTF-8 -t UTF-8 -x nfc "%~1" -o "~work.tmp"
tools\dos2unix "~work.tmp"
MOVE /Y "~work.tmp" "%~1" > NUL
EXIT /B 0

:readme
TYPE tools\readme.txt
ECHO.
PAUSE
EXIT /B
```

**tools\readme.txt**
```
You need to download the following tools:
- uconv (contained in the ICU package): https://github.com/unicode-org/icu/releases/latest
- dos2unix: https://dos2unix.sourceforge.io/
Extract the following files in the "\tools" folder:
	uconv.exe (and its dlls)
	dos2unix.exe
```

> Do not exceed 79 character on a line, or commands such as `config -h fullscreen`
won't be able to display the help properly, and might wrap the text or display blank
lines. The limit might be increased to 80 characters in the future, but
the implementation is not there yet.

Windows Notepad reports 79 characters as 80 columns, it starts with `line 1, column 1`.

**Comment by FeralChild64** [comment]
@Kappa971 OK, I have restored `dos2unix` - don't worry, mistakes do happen. I have also added your one-liner to generate current `en.lng`, but left the previous descriptions - not everyone likes such a long and complicated commands.

**Comment by Kappa971** [comment]
You can also add that you can use programs like `WinMerge` (Windows) or `Meld` (Linux) to compare an old `en.lng` file with a new one, to check for changes.

**Comment by FeralChild64** [comment]
@Kappa971 OK, I have added information about diff tools, also added some more information about suitable text editors.

File: contrib/resources/translations/README.md [hunk 17]
```diff
@@ -0,0 +1,98 @@
+
+# General translation workflow
+
+## Preparing file with English strings
+
+First, you need an up to date file with English strings. The one present in this
+directory is most likely outdated; to generate the current one either execute:
```
Comment (kcgen) [inline contrib/resources/translations/README.md:7]: (small suggestion)

Be sure you have the latest English messages file, `en.lng`, by running:

File: contrib/resources/translations/README.md [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+```
+dosbox -lang en -c "config -wl en.lng" -c "exit"
+```
+
+or do the following:
+
+- In the configuration file, go to the `[dosbox]` section and set`language = en`.
+- Run the most current version of DOSBox Staging.
+- Inside DOSBox execute the command `config -wl en.lng` (you can supply a host OS
+  path, instead of simply `en.lng`).
```
Comment (kcgen) [inline contrib/resources/translations/README.md:18]: These steps need to be done in order, so suggest replacing bullets with a sequence.

```markdown
1. In the configuration file, go to the `[dosbox]` section and set`language = en`.
1. Run the most current version of DOSBox Staging.
1. Inside DOSBox execute the command `config -wl en.lng` (you can supply a host OS
  path, instead of simply `en.lng`).
```

(the `1.` replaces the `-` and becomes a sequence at rendering time)

File: contrib/resources/translations/README.md [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+This will produce a file with the current strings to translate, use it as a
+reference.
+
+## Updating the translation
+
+To update translation just edit the relevant `*.lng` file with a text editor and save
+the file as UTF-8 encoded with Unix line endings.
```
Comment (kcgen) [inline contrib/resources/translations/README.md:26]: Small suggestion:

Update a given translation by editing the corresponding `*.lng` file with a text editor and saving it with UTF-8 encoding and Unix line endings.

File: contrib/resources/translations/README.md [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+to use the script to make sure no problem occurs.
+
+## String-specific tips
+
+### Format strings
+
+Many strings contain strangely looking sequences starting from the percentage sign,
+like %s or %02d. These are so-called _printf format strings_, used by C/C++
+programming languages; they are replaced with another string or number, using the
+specified format.
```
Comment (kcgen) [inline contrib/resources/translations/README.md:48]: :+1: ; very helpful!

File: contrib/resources/translations/normalize.bat [hunk 21]
```diff
@@ -0,0 +1,31 @@
+@ECHO OFF
```
Comment (kcgen) [inline contrib/resources/translations/normalize.bat:1]: Is it possible to add a license header?

```bat
REM SPDX-License-Identifier: GPL-2.0-or-later
REM
REM Copyright (C) 2022-2023  The DOSBox Staging Team
```

@Kappa971 , are you the author? Maybe this can be a separate commit with your as the commit author. Alternatively, if @FeralChild64 helped or made adjustments, use co-author: attribution in the commit.

Comment (kcgen) [reply contrib/resources/translations/normalize.bat:1]: Yes, I know this seems petty or unnecessary; this is more about eliminating ambiguity and unknowns.

**Review by kcgen — APPROVED** [review]
Just a couple small suggestions.

File: contrib/resources/translations/normalize.bat (hunk 21 above)
Comment (FeralChild64) [reply contrib/resources/translations/normalize.bat:1]: This is an old file, previously called encode.bat - it doesn't encode anything anymore, so I have renamed it. Unfortunately, there was enough changes that GIT stopped recognizing this as renamed file.

File: contrib/resources/translations/README.md (hunk 17 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: Changed, but I have kept a warning that the supplied file is most likely outdated.

File: contrib/resources/translations/README.md (hunk 18 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/README.md (hunk 19 above)
Comment (FeralChild64) [reply contrib/resources/translations/README.md]: OK, changed.

File: contrib/resources/translations/normalize.bat (hunk 21 above)
Comment (FeralChild64) [reply contrib/resources/translations/normalize.bat:1]: License header added. @kcgen - not sure how to proceed with copyright in this case; add a co-author? I would prefer to keep everything in one commit, or else we will have a non-working script in some point of GIT history.

Comment (kcgen) [reply contrib/resources/translations/normalize.bat:1]: ahh not a big deal; one commit is good.

**Comment by Kappa971** [comment]
```
display the help properly. Also test that that the `help /all` command properly
```
There are two "that" :)

**Review by Kappa971 — APPROVED** [review]
```
display the help properly. Also test that that the `help /all` command properly
```
There are two "that" :)

For the rest, you can go!

**Comment by FeralChild64** [comment]
> There are two "that" :)

Fixed, thanks.

**Comment by ThomasEricB** [comment]
I really like this, thanks @FeralChild64

**Comment by Kappa971** [comment]
I've observed that Notepad++ automatically adds tabs when needed. This causes some confusion in DOSBox, it doesn't seem to handle tabs well. The solution is to either use another text editor or have Notepad++ convert tabs to spaces (screenshot taken from a forum): 
scr1
Another option is to keep the tabs and convert them to spaces once you're done editing: `Edit > Space Operation > Convert TAB to Spaces`.
